tensorflow/core/transforms/remapper/pass.cc

```diff
@@ -91,18 +96,227 @@ class MatchMulSigmoid : public RewritePattern {
+  Value filter = contraction_op->getOperand(1);
+  Value bias = bias_add_op->getOperand(1);
+  operands.push_back(input);
+  operands.push_back(filter);
+  operands.push_back(bias);
+  llvm::append_range(operands, TFOp(contraction_op).getControlOperands());
+  llvm::append_range(operands, TFOp(bias_add_op).getControlOperands());
+
+  SmallVector<Location, 4> fused_locs{contraction_op->getLoc(),
+                                      bias_add_op->getLoc()};
+
+  NamedAttrList fused_attrs(contraction_op->getAttrs());
+  fused_attrs.set("fused_ops", rewriter.getStrArrayAttr({"BiasAdd"}));
+  fused_attrs.set("num_args", rewriter.getI32IntegerAttr(1));
+  // Default values for epsilon and leakyrelu_alpha
+  fused_attrs.set("epsilon", rewriter.getF32FloatAttr(0.0001));
+  fused_attrs.set("leakyrelu_alpha", rewriter.getF32FloatAttr(0.2));
+
+  OperationState state(rewriter.getFusedLoc(fused_locs), fused_op_name,
+                       operands, bias_add_op->getResultTypes(), fused_attrs);
+
+  if (auto new_op = rewriter.create(state)) {
+    return TFOp(new_op);
+  }
+
+  return failure();
+}
+
+// Contraction + BiasAdd
+// TODO(intel-tf): Support Contraction + {Add, AddV2} fusion in the case it has
+// similar semantic of contraction + BiasAdd
+class ContractionBiasAddRewriter : public RewriterBase {
+ public:
+  explicit ContractionBiasAddRewriter(OpPropertyHelper &helper)
+      : RewriterBase("tfg.BiasAdd", helper, PatternBenefit(1)) {}
+
+  // Constructor used by derived pattern rewritter class that may have
+  // different root operation name. Currently, pattern is
+  // matched from root op to its inputs.
+  explicit ContractionBiasAddRewriter(StringRef op_name,
+                                      OpPropertyHelper &helper,
+                                      PatternBenefit benefit)
+      : RewriterBase(op_name, helper, benefit) {}
+
+  using Pattern = ContractionBiasAdd;
+
+  bool matchPattern(Operation *op, Pattern &pattern) const {
+    // Not allowing control flow on BiasAdd
+    if (helper_.HasControlOperandsOrResultUsers(op)) return false;
+    Operation *contraction_op = op->getOperand(0).getDefiningOp();
+    if (!helper_.IsContraction(contraction_op) ||
+        helper_.HasControlOperandsOrResultUsers(contraction_op) ||
+        !helper_.HaveSameDataType(op, contraction_op) ||
+        !helper_.HasAtMostOneUserOfResult0(contraction_op))
+      return false;
+
+    pattern.contraction = contraction_op;
+    pattern.bias_add = op;
+    return true;
+  }
+
+  LogicalResult matchAndRewrite(Operation *op,
+                                PatternRewriter &rewriter) const override {
+    Pattern pattern;
+    if (!matchPattern(op, pattern)) return failure();
+    if (!helper_.IsDeviceCompatible(pattern)) return failure();
+    FailureOr<TFOp> fused_op = CreateContractionBiasAddOp(
+        rewriter, helper_, pattern.contraction, pattern.bias_add);
+    if (failed(fused_op)) return failure();
+    fused_op->setName(TFOp(op).nameAttr());
```

Comment:
I think you already have the name set in CreateContractionBiasAddOp (op->getAttrs() will include it). Besides, I would like to suggest that if you can prepare the attributes before operation creation, that'll be ideal. I know you may see some other cases which do post name setup but those are supposed to be improved.

Reply:
Refactored the function. Now we create OperationState before operation creation. Thanks!